### ISSUE TITLE
[Bug]:  基于chatglm2-6b 微调训练，报错

### ISSUE THREAD

**Issue by xieyu0214** [PaddlePaddle/PaddleNLP#6440]
### 软件环境

```Markdown
- paddlepaddle:
- paddlepaddle-gpu: 0.0.0.post117
- paddlenlp: 2.6.0rc0.post0

用的最新版代码
```

### 重复问题

- [X] I have searched the existing issues

### 错误描述

```Markdown
使用llm/chatglm/finetune_generation.py代码训练chatglm2-6b模型:

错误信息如下：

`Error: ../paddle/phi/kernels/funcs/gather.cu.h:67 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [2048] and greater than or equal to 0, but received [0]
Error: ../paddle/phi/kernels/funcs/gather.cu.h:67 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [2048] and greater than or equal to 0, but received [0]`


使用的配置为：
python -m paddle.distributed.launch --gpus "2,3" finetune_generation.py \
    --model_name_or_path THUDM/chatglm2-6b \
    --num_train_epochs 10 \
    --learning_rate 3e-5 \
    --warmup_steps 20 \
    --logging_steps 50 \
    --save_steps 100 \
    --save_total_limit 100 \
    --output_dir /chatglm/checkpoints/chatglm-6b-8-32-0719 \
    --src_length 4000 \
    --tgt_length 160 \
    --per_device_eval_batch_size 2 \
    --per_device_train_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --fp16 \
    --fp16_opt_level O2 \
    --recompute True \
    --do_train \
    --disable_tqdm True \
    --tensor_parallel_degree 2 \
    --lora True \
    --task_name_or_path /gen_dialogues/train_0718/


请问该怎么解决？
```

### 稳定复现步骤 & 代码

可以复现

**Comment by xieyu0214** [comment]
@sijunhe 请问是否支持了chatglm2的微调呢

**Comment by sijunhe** [comment]
> @sijunhe 请问是否支持了chatglm2的微调呢

你好，我这边尝试了一下双卡lora分布式没问题。请问你这个报错是训了一会才开始有的吗？

**Comment by xieyu0214** [comment]
> > @sijunhe 请问是否支持了chatglm2的微调呢
> 
> 你好，我这边尝试了一下双卡lora分布式没问题。请问你这个报错是训了一会才开始有的吗？

更新的最新的代码，重新测试，发现报错：
Error: ../paddle/phi/kernels/funcs/gather.cu.h:67 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [2048] and greater than or equal to 0, but received [0]

这个错误看起来是是不能设置--src_length 大于2048 ？？

参数配置如下：
python -m paddle.distributed.launch --gpus "0,1" finetune_generation.py \
    --model_name_or_path THUDM/chatglm2-6b \
    --num_train_epochs 3 \
    --learning_rate 3e-5 \
    --warmup_steps 20 \
    --logging_steps 50 \
    --save_steps 100 \
    --save_total_limit 100 \
    --output_dir /checkpoints/chatglm2-6b-8-32-0721-prefix \
    --src_length 4096 \
    --tgt_length 160 \
    --per_device_eval_batch_size 2 \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 32 \
    --fp16 \
    --fp16_opt_level O2 \
    --recompute True \
    --do_train \
    --disable_tqdm True \
    --tensor_parallel_degree 2 \
    --lora True \
    --task_name_or_path /assistant/gen_dialogues/train_0721_prefix/


看到有warning:
/home/xy/gitlabs/PaddleNLP/paddlenlp/transformers/tokenizer_utils_base.py:1894: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.

**Comment by xieyu0214** [comment]
> > @sijunhe 请问是否支持了chatglm2的微调呢
> 
> 你好，我这边尝试了一下双卡lora分布式没问题。请问你这个报错是训了一会才开始有的吗？

查了一下报错那条数据：
![image](https://github.com/PaddlePaddle/PaddleNLP/assets/13896613/4d55ea9d-27ac-465d-ba25-a8850ca1b64d)

看起来是不支持长度大于2048。 chatglm2-6b 长度已经扩充到8k了是吧？

**Comment by sijunhe** [comment]
> 看起来是不支持长度大于2048。 chatglm2-6b 长度已经扩充到8k了是吧？

感谢您的反馈。这块确实是我们漏加了一个小东西，#6534 修好了，应该后续就是32k的长度了

**Comment by xieyu0214** [comment]
> > 看起来是不支持长度大于2048。 chatglm2-6b 长度已经扩充到8k了是吧？
> 
> 感谢您的反馈。这块确实是我们漏加了一个小东西，#6534 修好了，应该后续就是32k的长度了

最新代码 基于chatglm2在具体项目上lora微调，训练了3个epoch，看起来在训练过程中loss也可以正常下降。但是预测时，发现效果很差（在之前的代码上，基于chatglm训练的模型效果还是可以的），所以猜测是代码中有bug的。现在不知道从何查起。。

你们有没有在具体项目上测试过最新代码呢?  是否验证过效果

**Comment by sijunhe** [comment]
> 看起来在训练过程中loss也可以正常下降。但是预测时，发现效果很差

1. 训练过程中的eval指标是正常的吗
2. 预测的脚本用的是哪个呢？

**Comment by xieyu0214** [comment]
> > 看起来在训练过程中loss也可以正常下降。但是预测时，发现效果很差
> 
> 1. 训练过程中的eval指标是正常的吗
> 2. 预测的脚本用的是哪个呢？

1. 为了省时间没有打开 --do_eval  
2. 用的是https://github.com/PaddlePaddle/PaddleNLP/blob/develop/llm/chatglm/predict_generation.py 这个脚本


       https://github.com/PaddlePaddle/PaddleNLP/blob/6e770820d0a838fa4826359840f9932b696ea40d/llm/chatglm/predict_generation.py#L118  
dedug时，上面这行的结果是：
![image](https://github.com/PaddlePaddle/PaddleNLP/assets/13896613/9536b90f-0028-42bf-ace1-c7efdbe64094)

看起来就不太对。。

**Comment by sijunhe** [comment]
> 看起来就不太对

怎么个不太对法呢？我们这边做的是left padding, 模型输入符合我的预期。
可以尝试推理原有模型版本，看下是不是同样的有问题，同时也可以考虑训练中打开eval

**Comment by xieyu0214** [comment]
> > 看起来就不太对
> 
> 怎么个不太对法呢？我们这边做的是left padding, 模型输入符合我的预期。 可以尝试推理原有模型版本，看下是不是同样的有问题，同时也可以考虑训练中打开eval

哦对 left padding。 `可以尝试推理原有模型版本` 这个 原有模型是什么意思呢？

**Comment by sijunhe** [comment]
> 哦对 left padding。 `可以尝试推理原有模型版本` 这个 原有模型是什么意思呢？

就是直接推理预训练的chatglm2模型，看看输出正常不。

**Comment by xieyu0214** [comment]
> > 哦对 left padding。 `可以尝试推理原有模型版本` 这个 原有模型是什么意思呢？
> 
> 就是直接推理预训练的chatglm2模型，看看输出正常不。

cases = ["你好", "晚上睡不着应该怎么办"]
'result': ['，我是 ChatGLM2-6B， 一个人工智能助手。我背后使用的模型是 GLM2-6B， 是一种大型语言模型， 具有超过 3000 亿参数，支持多种任务。', '？ 首先，我建议在晚上避免使用电子设备，如手机、电脑等，因为这些设备会发出蓝光，可能会干扰睡眠。其次，在睡觉前可以尝试放松身心，比如听轻柔的音乐、做瑜伽、冥想等，这些活动可以帮助放松身体和思维，有助于睡眠。另外，保持室内安静 、舒适也是有助于睡眠的。最后，如果仍然无法入睡，建议在白天多进行一些运动，但不要在晚上进行剧烈的运动，以免影响睡眠。']

测试了官网的两个case，结果看起来不太对


（好像在其他代码中遇到类似的情况，当时的问题是代码中EOS的设置不对。）

**Comment by sijunhe** [comment]
> 
这个结果和torch那边是对齐的呢。只是transformers的generate会把input也一起合进去，我们这边只输出output

**Comment by xieyu0214** [comment]
> > 
> 
> 这个结果和torch那边是对齐的呢。只是transformers的generate会把input也一起合进去，我们这边只输出output

（https://huggingface.co/THUDM/chatglm2-6b）torch的结果不是这个吗？
![image](https://github.com/PaddlePaddle/PaddleNLP/assets/13896613/a2eed849-cfdd-44a6-8e3c-877923dcf388)
